working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,612 @@
+
+### !end-callout
+
+### !callout-info
+
+## Running the Project and Tests
+
+In this branch, the startup code in `app/__init__.py` has been updated to read its connection string from an environment variable. If we are working in a new cloned repo and don't have a `.env`, the following file can be used:
+
+<br />
+
+<details>
+<summary><code>.env</code></summary>
+
+```bash
+SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_development
+SQLALCHEMY_TEST_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_test
+```
+
+</details>
+
+<br />
+
+Be sure that the `hello_books_development` and `hello_books_test` databases exist, and that the migrations have been run to create the tables in the development database. If there are already tables there, dropping and recreating the database will allow us to re-run the migrations if we're unsure of the current state of the database.
+
+### !end-callout
+
+### Planning the `validate_book` Refactor
+
+Our steps to plan a refactor with Copilot stay the same: we first need to identify our dependencies, then check if we have tests for code that will be affected. We will be refactoring the `validate_book` function in the `routes.py` file, so we should search the project for where the function is used to remind ourselves of our dependencies.
+
+This is a scenario where Copilot shouldn't be used, since Copilot has access to open files, but not all of the files in your project. Indeed, if we ask Copilot for help, it will warn us about this lack of access and give us suggestions on how to search the project for dependencies.
+
+<br />
+
+<details>
+  <summary>
+    If curious, feel free to try asking Copilot to identify dependencies for <code>validate_book</code> or expand this section to see our prompt and what Copilot suggested.
+  </summary>
+
+  **Prompt:**
+  > What code in this project is dependent on the validate_book function?
+
+  **Copilot's Response:**
+  > As an AI, I don't have access to your entire project's codebase, so I can't directly identify all the places where `validate_book` function is used. However, typically, such a function would be used in routes that require a book ID as a parameter.
+  >  
+  > For example, you might have routes like `@books_bp.route("/<book_id>", methods=["GET", "PUT", "DELETE"])` where you would use `validate_book(book_id)` to ensure the provided book ID is valid before proceeding with the GET, PUT, or DELETE operation.
+  >  
+  > Please search your codebase for `validate_book` to find all its usages. In Visual Studio Code, you can do this by pressing `Ctrl+Shift+F` and typing `validate_book` in the search bar.
+</details>
+
+If we search our project, we have 4 dependencies:
```

Comment:
We have 4 dependencies whether or not we search. Suggest strengthening the action and making the outcome dependent on the action (less important due to removing the condition)
```suggestion
Searching our project, we find 4 dependencies:
```